File: test/integration/ContainerViewsMain.go
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//
//
// The C probe's reading of a container holding a view, in Go: the view is a slice of the buffer.
//
//===----------------------------------------------------------------------===//

package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"math"
	"os"

	aliasable "llvmdsdl_generated/fixtures_aliasable/vendor"
	views "llvmdsdl_generated/fixtures_views/vendor"
)

var failures = 0

func check(what string, ok bool) {
	v := "ok"
	if !ok {
		v = "FAILED"
		failures++
	}
	fmt.Printf("  %-44s %s\n", what, v)
}

func main() {
	wire := make([]byte, 41)
	binary.LittleEndian.PutUint32(wire, 0x11223344)
	for i, v := range []float32{1.5, -2.25, 3.0, 4.0, 5.5, 6.75} {
		binary.LittleEndian.PutUint32(wire[4+i*4:], math.Float32bits(v))
	}
	for i, v := range []float32{7.0, 8.0, 9.0} {
		binary.LittleEndian.PutUint32(wire[28+i*4:], math.Float32bits(v))
	}
	wire[40] = 0x5A
	var frame views.Frame
	n, err := frame.Deserialize(wire)
	check("deserialise accepted", err == nil && n == 41)
	check("sequence decoded", frame.Sequence == 0x11223344)
	check("view points into the buffer", len(frame.Pose) == 24 && &frame.Pose[0] == &wire[4])
	check("orientation.y read through the view", aliasable.Vec3GetY(aliasable.PoseGetOrientation(frame.Pose)) == 5.5)
	check("velocity decoded, status after the view", frame.Velocity.Y == 8.0 && frame.Status == 0x5A)
	out := bytes.Repeat([]byte{0xEE}, 64)
	n, err = frame.Serialize(out)
	check("serialise reproduces the wire", err == nil && n == 41 && bytes.Equal(out[:41], wire))
	var short views.Frame
	_, err = short.Deserialize(wire[:16])
	check("short deserialise accepted", err == nil)
	check("short view holds what was there", len(short.Pose) == 12 && &short.Pose[0] == &wire[4])
	o := aliasable.PoseGetOrientation(short.Pose)
	check("missing orientation reads as zero", len(o) == 0 && aliasable.Vec3GetY(o) == 0 && short.Status == 0)
	out = bytes.Repeat([]byte{0xEE}, 64)
	_, err = short.Serialize(out)
	check("short view serialises zero-filled", err == nil && bytes.Equal(out[4:16], wire[4:16]) && bytes.Equal(out[16:28], make([]byte, 12)))
	var fresh views.Frame
	check("fresh object holds an empty view", len(fresh.Pose) == 0)
	out = bytes.Repeat([]byte{0xEE}, 64)
	_, err = fresh.Serialize(out)
	check("empty view serialises as zeros", err == nil && bytes.Equal(out[4:28], make([]byte, 24)))
	trackWire := make([]byte, 99)
	trackWire[0] = 0x07
	poses := [][]float32{{1, 2, 3, 4, 5, 6}, {10, 20, 30, 40, 50, 60}, {-1, -2, -3, -4, -5, -6}, {0.5, 1.5, 2.5, 3.5, 4.5, 5.5}}
	for p, at := range []int{1, 25, 50, 74} {
		for i, v := range poses[p] {
			binary.LittleEndian.PutUint32(trackWire[at+i*4:], math.Float32bits(v))
		}
	}
	trackWire[49] = 2
	trackWire[98] = 0x3C
	var track views.Track
	n, err = track.Deserialize(trackWire)
	check("track deserialise accepted", err == nil && n == 99)
	check("pair elements are views into the buffer", len(track.Pair[0]) == 24 && &track.Pair[0][0] == &trackWire[1] && len(track.Pair[1]) == 24 && &track.Pair[1][0] == &trackWire[25])
	check("trail keeps its count, elements are views", len(track.Trail) == 2 && len(track.Trail[1]) == 24 && &track.Trail[1][0] == &trackWire[74])
	check("orientation.y read through pair[1]", aliasable.Vec3GetY(aliasable.PoseGetOrientation(track.Pair[1])) == 50.0 && track.Kind == 0x07 && track.Status == 0x3C)
	out = bytes.Repeat([]byte{0xEE}, 128)
	n, err = track.Serialize(out)
	check("track serialise reproduces the wire", err == nil && n == 99 && bytes.Equal(out[:99], trackWire))
	var shortTrack views.Track
	_, err = shortTrack.Deserialize(trackWire[:37])
	check("short track: pair[1] short, trail empty", err == nil && len(shortTrack.Pair[1]) == 12 && &shortTrack.Pair[1][0] == &trackWire[25] && len(shortTrack.Trail) == 0 && shortTrack.Status == 0)
	out = bytes.Repeat([]byte{0xEE}, 128)
	n, err = shortTrack.Serialize(out)
	check("short element serialises zero-filled", err == nil && n == 51 && bytes.Equal(out[25:37], trackWire[25:37]) && bytes.Equal(out[37:51], make([]byte, 14)))
	var freshTrack views.Track
	check("fresh object holds empty element views", len(freshTrack.Pair[0]) == 0 && len(freshTrack.Pair[1]) == 0 && len(freshTrack.Trail) == 0)
	out = bytes.Repeat([]byte{0xEE}, 128)
	n, err = freshTrack.Serialize(out)
	check("empty element views serialise as zeros", err == nil && n == 51 && bytes.Equal(out[1:51], make([]byte, 50)))
	leadWire := make([]byte, 25)
	for i, v := range []float32{0.25, -0.5, 0.75, -1.0, 1.25, -1.5} {
		binary.LittleEndian.PutUint32(leadWire[i*4:], math.Float32bits(v))
	}
	leadWire[24] = 0xA5
	var lead views.Leading
	n, err = lead.Deserialize(leadWire)
	check("leading deserialise accepted", err == nil && n == 25)
	check("leading view is the buffer itself", len(lead.Pose) == 24 && &lead.Pose[0] == &leadWire[0])
	check("orientation.y read through the leading view", aliasable.Vec3GetY(aliasable.PoseGetOrientation(lead.Pose)) == 1.25 && lead.Status == 0xA5)
	out = bytes.Repeat([]byte{0xEE}, 64)
	n, err = lead.Serialize(out)
	check("leading serialise reproduces the wire", err == nil && n == 25 && bytes.Equal(out[:25], leadWire))
	var shortLead views.Leading
	_, err = shortLead.Deserialize(leadWire[:12])
	check("short leading view holds what was there", err == nil && len(shortLead.Pose) == 12 && &shortLead.Pose[0] == &leadWire[0] && shortLead.Status == 0)
	out = bytes.Repeat([]byte{0xEE}, 64)
	n, err = shortLead.Serialize(out)
	check("short leading view serialises zero-filled", err == nil && n == 25 && bytes.Equal(out[:12], leadWire[:12]) && bytes.Equal(out[12:25], make([]byte, 13)))
	var emptyLead views.Leading
	n, err = emptyLead.Deserialize(nil)
	o = aliasable.PoseGetOrientation(emptyLead.Pose)
	check("nil buffer leaves an empty leading view", err == nil && n == 0 && len(emptyLead.Pose) == 0 && emptyLead.Status == 0 && len(o) == 0 && aliasable.Vec3GetY(o) == 0)
	var freshLead views.Leading
	check("fresh object holds an empty leading view", len(freshLead.Pose) == 0)
	out = bytes.Repeat([]byte{0xEE}, 64)
	n, err = freshLead.Serialize(out)
	check("empty leading view serialises as zeros", err == nil && n == 25 && bytes.Equal(out[:25], make([]byte, 25)))
	verdict := "ok"
	if failures != 0 {
		verdict = "FAILED"
	}
	fmt.Printf("container-views Go: %s\n", verdict)
	if failures != 0 {
		os.Exit(1)
	}
}

File: test/integration/ContainerViewsMain.rs
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//
//
// The C probe's reading of a container holding a view, in Rust: the holder borrows the buffer for its lifetime.
//
//===----------------------------------------------------------------------===//

use llvmdsdl_generated::fixtures_aliasable::vendor::pose_1_0::fixtures_aliasable_vendor_Pose as Pose;
use llvmdsdl_generated::fixtures_aliasable::vendor::vec3_1_0::fixtures_aliasable_vendor_Vec3 as Vec3;
use llvmdsdl_generated::fixtures_views::vendor::frame_1_0::fixtures_views_vendor_Frame as Frame;
use llvmdsdl_generated::fixtures_views::vendor::track_1_0::fixtures_views_vendor_Track as Track;
fn main() {
    let mut failures = 0;
    let mut check = |what: &str, ok: bool| { println!("  {:<44} {}", what, if ok { "ok" } else { "FAILED" }); if !ok { failures += 1; } };
    let mut wire = [0u8; 41];
    wire[0..4].copy_from_slice(&0x11223344u32.to_le_bytes());
    for (i, v) in [1.5f32, -2.25, 3.0, 4.0, 5.5, 6.75].iter().enumerate() { wire[4 + i * 4..8 + i * 4].copy_from_slice(&v.to_le_bytes()); }
    for (i, v) in [7.0f32, 8.0, 9.0].iter().enumerate() { wire[28 + i * 4..32 + i * 4].copy_from_slice(&v.to_le_bytes()); }
    wire[40] = 0x5A;
    let mut frame = Frame::default();
    check("deserialise accepted", frame.deserialize(&wire) == Ok(41));
    check("sequence decoded", frame.sequence == 0x11223344);
    check("view points into the buffer", frame.pose.as_ptr() == wire[4..].as_ptr() && frame.pose.len() == 24);
    check("orientation.y read through the view", Vec3::get_y(Pose::get_orientation(frame.pose)) == 5.5);
    check("velocity decoded, status after the view", frame.velocity.y == 8.0 && frame.status == 0x5A);
    let mut out = [0xEEu8; 64];
    check("serialise reproduces the wire", frame.serialize(&mut out) == Ok(41) && out[..41] == wire);
    let mut short = Frame::default();
    check("short deserialise accepted", short.deserialize(&wire[..16]).is_ok());
    check("short view holds what was there", short.pose.len() == 12 && short.pose.as_ptr() == wire[4..].as_ptr());
    let o = Pose::get_orientation(short.pose);
    check("missing orientation reads as zero", o.is_empty() && Vec3::get_y(o) == 0.0 && short.status == 0);
    let mut out = [0xEEu8; 64];
    check("short view serialises zero-filled", short.serialize(&mut out).is_ok() && out[4..16] == wire[4..16] && out[16..28].iter().all(|&b| b == 0));
    let fresh = Frame::default();
    check("fresh object holds an empty view", fresh.pose.is_empty());
    let mut out = [0xEEu8; 64];
    check("empty view serialises as zeros", fresh.serialize(&mut out).is_ok() && out[4..28].iter().all(|&b| b == 0));
    let mut track_wire = [0u8; 99];
    track_wire[0] = 0x07;
    let poses: [[f32; 6]; 4] = [[1.0, 2.0, 3.0, 4.0, 5.0, 6.0], [10.0, 20.0, 30.0, 40.0, 50.0, 60.0], [-1.0, -2.0, -3.0, -4.0, -5.0, -6.0], [0.5, 1.5, 2.5, 3.5, 4.5, 5.5]];
    for (at, pose) in [1usize, 25, 50, 74].iter().zip(poses.iter()) { for (i, v) in pose.iter().enumerate() { track_wire[at + i * 4..at + i * 4 + 4].copy_from_slice(&v.to_le_bytes()); } }
    track_wire[49] = 2;
    track_wire[98] = 0x3C;
    let mut track = Track::default();
    check("track deserialise accepted", track.deserialize(&track_wire) == Ok(99));
    check("pair elements are views into the buffer", track.pair[0].as_ptr() == track_wire[1..].as_ptr() && track.pair[0].len() == 24 && track.pair[1].as_ptr() == track_wire[25..].as_ptr() && track.pair[1].len() == 24);
    check("trail keeps its count, elements are views", track.trail.len() == 2 && track.trail[1].as_ptr() == track_wire[74..].as_ptr() && track.trail[1].len() == 24);
    check("orientation.y read through pair[1]", Vec3::get_y(Pose::get_orientation(track.pair[1])) == 50.0 && track.kind == 0x07 && track.status == 0x3C);
    let mut out = [0xEEu8; 128];
    check("track serialise reproduces the wire", track.serialize(&mut out) == Ok(99) && out[..99] == track_wire);
    let mut short_track = Track::default();
    check("short track: pair[1] short, trail empty", short_track.deserialize(&track_wire[..37]).is_ok() && short_track.pair[1].as_ptr() == track_wire[25..].as_ptr() && short_track.pair[1].len() == 12 && short_track.trail.len() == 0 && short_track.status == 0);
    let mut out = [0xEEu8; 128];
    check("short element serialises zero-filled", short_track.serialize(&mut out) == Ok(51) && out[25..37] == track_wire[25..37] && out[37..51].iter().all(|&b| b == 0));
    let fresh_track = Track::default();
    check("fresh object holds empty element views", fresh_track.pair.iter().all(|p| p.is_empty()) && fresh_track.trail.len() == 0);
    let mut out = [0xEEu8; 128];
    check("empty element views serialise as zeros", fresh_track.serialize(&mut out) == Ok(51) && out[1..51].iter().all(|&b| b == 0));
    println!("container-views Rust: {}", if failures == 0 { "ok" } else { "FAILED" });
    std::process::exit(if failures == 0 { 0 } else { 1 });
}
